### PR TITLE
[refactor]: shared attention infra additions for QAT-compat (Attn-QAT 5/12)

### PR DESCRIPTION
**Pull request by SolitaryThinker**
Slice 5 of 12 in the PR #1225 decomposition. **Tier 2 — backward-compat additions.**

## Summary

Shared attention-backend infra additions to support the QAT
(Quantization-Aware Training) backends added in slice 4 (PR #1358, merged
as `fda02036b`). All changes are backward-compatible: existing call sites
keep working; the additions provide a uniform metadata interface and a
centralized varlen-flash-attn fallback so future QAT-aware backends
(slices 6+) can plug in without re-declaring shared state per file.

## Files

| File | Δ | What it adds |
|---|---|---|
| `fastvideo/attention/backends/abstract.py` | +6/-2 | Moves `VSA_sparsity: float = 0.0` (kw-only) into base `AttentionMetadata`; adds `__getattr__` with clean `AttributeError`; loosens `AttentionMetadataBuilder.build` kwargs type to `Any`. |
| `fastvideo/attention/backends/bsa_attn.py` | +3/-5 | Collapses double-fallback flash-attn import to a single import from the now-centralized resolver. |
| `fastvideo/attention/backends/video_sparse_attn.py` | +2/-3 | Drops the redundant `VSA_sparsity` field from `VideoSparseAttentionMetadata` (now inherited from base); adds `-> None` annotations. |
| `fastvideo/attention/backends/vmoba.py` | +8/-4 | Adds `-> None` annotations; tightens `device: torch.device \| None = None`; narrows `attn_metadata` to `VideoMobaAttentionMetadata`; adds explicit guard + `else: raise ValueError` for the moba_layer chunk-selection switch; pre-declares `moba_chunk_size` for mypy. |
| `fastvideo/attention/layer.py` | +2 | Adds `attention_mask: torch.Tensor \| None = None` to `DistributedAttention_VSA.forward`. |
| `fastvideo/attention/utils/flash_attn_no_pad.py` | +47/-33 | Centralizes the varlen-flash-attn fallback chain (`flash_attn_cute` → `flash_attn_interface` → `flash_attn`) into `_resolve_flash_attn_varlen_func()`; adds type annotations to every public function. |

### Files in PR #1225 considered but NOT applied

- `fastvideo/attention/backends/sage_attn3.py` — the source-SHA diff shrinks `get_supported_head_sizes()` from `[64, 128, 256]` to `[64, 128]`. That removal is unrelated to QAT-compat (head_size=256 has been supported since the original SAGE3 backend landed in #815) and is treated as an obsolete edit from the source branch's history. Not applied; current `main`'s behavior is preserved.

## Stacking

- Base: `main` (slice 4 / PR #1358 merged at `fda02036b`)
- No stack dependency — this is a clean linear PR off `main`.

## Provenance

Files extracted from PR #1225 (https://github.com/hao-ai-lab/FastVideo/pull/1225) at source SHA `3f818d0fc532ec6494b465967d5f485150917d0c` and audited for newest-API compat. `abstract.py`, `vmoba.py`, `layer.py`, and `flash_attn_no_pad.py` were applied directly (main had not diverged from the source's merge-base for those files). `bsa_attn.py` and `video_sparse_attn.py` required a 3-way merge against current `main`; both merged cleanly with no manual conflict resolution required.

## Test plan

Two new CPU-only smoke test modules under `fastvideo/tests/attention/`:

- `test_attention_metadata_base.py` (8 tests, all pass): `VSA_sparsity` default 0.0, kw-only enforcement, subclass inheritance, subclass override via kw, `__getattr__` raises `AttributeError` with stable message, `asdict_zerocopy` includes/skips `VSA_sparsity` correctly.
- `test_flash_attn_no_pad_resolver.py` (3 tests): fallback-chain coverage for the centralized resolver (cute → interface → flash_attn). Skipped at module level when `flash_attn` is not installed.

Pre-commit gate (`yapf` + `ruff` + `codespell` + `mypy`) passes on all 8 changed/new files. Local run:

```
================== 8 passed, 1 skipped, 14 warnings in 0.02s ===================
```

## Sequence

Attn-QAT-Stack: 5/12. Earlier slices: 1/12 (`#?`), 2/12, 3/12 staged; 4/12 merged as PR #1358. Out of scope for this slice: linear/MLP FP4 path additions and downstream QAT-aware backend wiring (later slices).

### CODE REVIEW

File: fastvideo/attention/layer.py [hunk 1]
```diff
@@ -158,6 +158,7 @@ def forward(
         replicated_v: torch.Tensor | None = None,
         gate_compress: torch.Tensor | None = None,
         freqs_cis: tuple[torch.Tensor, torch.Tensor] | None = None,
+        attention_mask: torch.Tensor | None = None,
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/layer.py]: Should be addressed by 5fef55c1e — I dropped the unused `attention_mask` parameter (and its docstring line) on the same self-review pass. Slice 5 doesn't activate masking logic, so adding the kwarg without wiring it through was misleading. Let me know if you still see it.

File: fastvideo/attention/backends/abstract.py [hunk 2]
```diff
@@ -53,6 +53,10 @@ class AttentionMetadata:
     """Attention metadata for prefill and decode batched together."""
     # Current step of diffusion process
     current_timestep: int
+    VSA_sparsity: float = field(default=0.0, kw_only=True)
+
+    def __getattr__(self, name: str) -> Any:
+        raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/abstract.py:59]: I'd push back on this one — the `__getattr__` provides a stable error message format (`'{type}' object has no attribute '{name}'`) regardless of whether the missing access is on a dataclass field or an arbitrary lookup. The new test `test_attention_metadata_base.py::test_getattr_raises_attribute_error` asserts on that exact message shape, so removing the override would silently change the contract. Happy to revisit if you have a counterexample where Python's default message would suffice.
